### PR TITLE
feat(devcontainer): devcontainer and easy up robot

### PR DESCRIPTION
**Pull request by y3rsh**
# Overview

- Add devcontainer
- Add dev container to GitHub registry so that it builds faster
- Add `make` commands to stand up robots in one command
  - Example: `make ot2` or `make ot3`
- Docs for the one command robot
- Docs on using the devcontainer locally for more nuanced emulations.

### CODE REVIEW

**Comment by DerekMaggio** [comment]
> Where `make setup` ?

I would put it in the top level Makefile under `dev-container` or `vs-code-container` or something like that

File: emulation_system/Makefile [hunk 1]
```diff
@@ -1,7 +1,12 @@
+.PHONY: update
+update:
+	poetry env use 3.10
+	poetry update
+	poetry install
+
 .PHONY: setup
```
Comment (y3rsh) [inline emulation_system/Makefile:7]: @DerekMaggio this change cool?

File: Makefile [hunk 2]
```diff
@@ -260,3 +260,29 @@ format:
 .PHONY: test
 test:
 	$(MAKE) -C $(EMULATION_SYSTEM_DIR) test
+
+RUN_COMMAND := $(if $(CI),remove-build-run-detached,remove-build-run)
```
Comment (y3rsh) [inline Makefile:264]: - [ ] @y3rsh remove this.

File: Makefile [hunk 3]
```diff
@@ -5,7 +5,7 @@ SUB = {SUB}
 EMULATION_SYSTEM_CMD := (cd ./emulation_system && poetry run python main.py emulation-system {SUB} -)
 DEV_EMULATION_SYSTEM_CMD := (cd ./emulation_system && poetry run python main.py emulation-system --dev {SUB} -)
 REMOTE_ONLY_EMULATION_SYSTEM_CMD := (cd ./emulation_system && poetry run python main.py emulation-system {SUB} - --remote-only)
-COMPOSE_RUN_COMMAND := docker-compose -f - up
+COMPOSE_RUN_COMMAND := docker-compose -f - up --remove-orphans
```
Comment (y3rsh) [inline Makefile:8]: @DerekMaggio this ok?

File: .github/workflows/build-devcontainer.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
+name: "build-devcontainer"
+on: # Make sure the devcontainer changes are pushed on changes to main
+  push:
+    branches:
+      - main
+      - "233-devcontainer" # remove before merge
```
Comment (y3rsh) [inline .github/workflows/build-devcontainer.yaml:6]: - [ ] remove this line before merge

```suggestion
```

**Comment by y3rsh** [comment]
- [ ] @y3rsh create docs

File: emulation_system/Makefile (hunk 1 above)
Comment (DerekMaggio) [reply emulation_system/Makefile:7]: Yes sir

File: Makefile (hunk 3 above)
Comment (DerekMaggio) [reply Makefile:8]: Yeah, I don't see why not. Should just make everything start cleaner. 
If it ends up causing problems down the line we can just remove it.

File: .github/workflows/build-devcontainer.yaml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build and run dev container task
+        uses: devcontainers/ci@v0.2
+        with:
+          imageName: ghcr.io/opentrons/opentrons-emulation-devcontainer
+          runCmd: |
+            python --version
```
Comment (DerekMaggio) [inline .github/workflows/build-devcontainer.yaml:31]: What is this `runCmd` for?

File: scripts/docker_convenience_scripts/in_docker.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import os
```
Comment (DerekMaggio) [inline scripts/docker_convenience_scripts/in_docker.py:1]: What is this?

**Review by DerekMaggio — COMMENTED** [review]
Everything looks good!
Just some questions I have before approving.

File: .github/workflows/build-devcontainer.yaml (hunk 5 above)
Comment (y3rsh) [reply .github/workflows/build-devcontainer.yaml:31]: [That is the parameter of what command to run **in** the container once the container is up.  Here we don't need to run anything, so I am printing python version.](https://github.com/devcontainers/ci/blob/main/docs/github-action.md#inputs)

You could run `make setup` then `make lint` and do your lint check.

This part is still a bit clunky...  It is good for simple things but more complex things would not be fun at the current maturity.

File: scripts/docker_convenience_scripts/in_docker.py (hunk 6 above)
Comment (y3rsh) [reply scripts/docker_convenience_scripts/in_docker.py:1]: Detecting if you are running inside a docker container is tricky.  This script returns a string of `true` or `false`
We are using the output to determine the hostname.  When you are in a container (docker-from-docker) and then spin up container(s) on the host, like we are doing in the devcontainer, you cannot get to it from `localhost` from inside the devcontainer.  You need  `host.docker.internal`.  BUT the app on the host system would be able to resolve `localhost`.

**Comment by y3rsh** [comment]
Video for readme
https://user-images.githubusercontent.com/502770/194157913-316e8115-2ff4-4658-a274-8b22f2bdef5c.mp4

**Comment by y3rsh** [comment]
Dev container build done image.
image
image

**Comment by y3rsh** [comment]
Set localhost in app

https://user-images.githubusercontent.com/502770/194162290-48c370f9-2b1b-4a75-995b-d6f1dc2e39bb.mp4

File: Makefile [hunk 7]
```diff
@@ -260,3 +260,29 @@ format:
...
+	$(MAKE) setup
+	cp configuration_ci.json configuration.json
+	$(MAKE) check-remote-only file_path="$(OT3CONFIG)"
+	$(MAKE) $(RUN_COMMAND) file_path="$(OT3CONFIG)"
+
+ROBOT_HOST := $(if $(shell python ./scripts/docker_convenience_scripts/in_docker.py),host.docker.internal,localhost)
+
+.PHONY: check-robot
+check-robot:
+	curl -s --location --request GET 'http://$(ROBOT_HOST):31950/modules' --header 'opentrons-version: *' | json_pp -json_opt pretty,canonical
```
[the remark this answers is not in the document]
Comment (y3rsh) [reply Makefile]: - Added the docs.
- Yes, useful samples as of now.  Need 
  - ot3 with modules when we get there
  - ot2 switch in Gen2 Thermocycler in the all_modules sample
- a user would have to have make on thier host system
  - not sure if possible
  - just as easy to open a new window of vs code and use the command palette, I can make a video of that.

**Review by DerekMaggio — APPROVED** [review]
Looks fantastic to me!
So excited for this!

File: .github/workflows/build-devcontainer.yaml [hunk 8]
```diff
@@ -1,9 +1,9 @@
 name: "build-devcontainer"
 on: # Make sure the devcontainer changes are pushed on changes to main
   push:
-    branches:
```
Comment (DerekMaggio) [inline .github/workflows/build-devcontainer.yaml]: Why are we not building against main anymore?

**Review by DerekMaggio — CHANGES_REQUESTED** [review]

**Pull request merged**